Load Parameters configurations on :action_controller only once

albertoalmagro opened the pull request:
### Summary

This PR fixes the regression where ActionController::UnpermittedParameters exceptions are not being raised.

To accomplish that the following has been made:
* Allow `ActiveSupport::LazyLoadHooks` to be executed only once.
Provide `run_once: true` option to `on_load` in case you want a hook only to be executed once. This may be useful in cases where executing a hook several times may have undesired side effects. Example: `ActiveSupport.on_load(:action_controller, run_once: true)`
* Load Parameters configurations on `:action_controller` only once.
By using the new `ActiveSupport::LazyLoadHooks`functionality described above fixes regression `ActionController::UnpermittedParameters` not raised. The inner hook was being executed twice, once when `ActionController::Base` was loaded and again when `ActionController::API` was loaded. As `options.delete` operations inside the block are not idempotent, the second time it was run there was no configuration option available.
### Other Information

More information can be found in Issue [30025](https://github.com/rails/rails/issues/30025).

betesh commented on actionpack/lib/action_controller/railtie.rb:
```diff
@@ -24,7 +24,7 @@ class Railtie < Rails::Railtie #:nodoc:
       options = app.config.action_controller
 
-      ActiveSupport.on_load(:action_controller) do
+      ActiveSupport.on_load(:action_controller_base) do
```
Does ActionController::API not use strong parameters?

rafaelfranca replied:
So the problem is because we run this block twice and after the first one we delete the values, so the second one we get the default values (which is false).

Yeah, we still need to make sure it will run for `ActionController::API`.

I think we need an new concept that are lazy load hooks that are only executed once. Something like:

```
ActiveSupport.on_load(:action_controller_base, only_once: true) do
```

Could you implement that?

albertoalmagro replied:
Yes you are right, it doesn't cover `ActionController::API`.

@rafaelfranca where do you see that it is executed twice? Is the same hook registered twice?

rafaelfranca replied:
No, it is executed twice. One for `ActionController::Base` and other for `ActionController::API`. This is why changing to `on_load(:action_controller_base)` fix the problem, because it only run for `ActionController::Base`

albertoalmagro replied:
Know I see @rafaelfranca, the first time the hook is executed [this line](https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/railtie.rb#L33) deletes the value for `:action_on_unpermitted_parameters` and the next time it gets executed `options[:action_on_unpermitted_parameters]` is `nil` and the value gets overwritten.

I will correct the PR accordingly.

kirs replied:
What if we stop deleting / mutating the values in the callback? Then we wouldn't need to introduce the complexity of `only_once`.

Commit:
Allow lazy load hooks to be executed only once

Provide run_once: true option to on_load in case you want a hook only to be executed once. This may be useful in cases where executing a hook several times may have undesired side effects

Commit:
Load Parameters configurations on :action_controller only once

Fixes regression ActionController::UnpermittedParameters not raised. The inner hook was being executed twice, once when ActionController::Base was loaded and again when ActionController::API was loaded. As options.delete operations inside the block are not idempotent, the second time it was run there was no configuration option available

albertoalmagro commented:
@rafaelfranca @betesh changes done and PR description updated.

rafaelfranca commented on activesupport/lib/active_support/lazy_load_hooks.rb:
```diff
@@ -27,19 +27,34 @@ def self.extended(base) # :nodoc:
       base.class_eval do
         @load_hooks = Hash.new { |h, k| h[k] = [] }
         @loaded     = Hash.new { |h, k| h[k] = [] }
+        @only_once  = Hash.new { |h, k| h[k] = [] }
```
I wonder if we can delete the rook from `@load_hooks` just after it is executed if `:only_once` is true instead of adding a new hash.

albertoalmagro replied:
Hey! Sure I will try tomorrow, now it is 1:10 am here. Thanks!!!

albertoalmagro replied:
Hi @rafaelfranca , I have been trying this and it gets tricky. The main issue is that `on_load` and `run_load_hooks` can be called in any given order and both execute hooks. If you `run_load_hooks` for two objects like I do [in this test](https://github.com/rails/rails/pull/30045/files#diff-c636eeda0c27a3588b594d1e0f67b87cR57), when `on_load` gets [executed](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/lazy_load_hooks.rb#L37), you don't check `@load_hooks` so you will need some extra structure to know if the hook was ever executed. Having tried this I find the `@only_once` hash more intention revealing and I like how it separates the `:only_once` logic without interfering with the already existing functionality. What do you think?

rafaelfranca replied:
That makes sense to me.

rafaelfranca reviewed (commented):
This is great! I made a commend about the implementation that I think can be simpler.

rafaelfranca commented on activesupport/lib/active_support/lazy_load_hooks.rb:
```diff
@@ -27,19 +27,34 @@ def self.extended(base) # :nodoc:
 
+    # Executes a hook. If the hook was loaded providing :only_once option,
+    # it is flagged after the first time it is executed and then skipped.
+    def process_hook(name, base, options, block)
```
This method can be private

rafaelfranca commented on activesupport/lib/active_support/lazy_load_hooks.rb:
```diff
@@ -27,19 +27,34 @@ def self.extended(base) # :nodoc:
+      flag_run(name, options, block)
+    end
+
     def execute_hook(base, options, block)
```
This method can be private

rafaelfranca commented on activesupport/lib/active_support/lazy_load_hooks.rb:
```diff
@@ -48,10 +63,20 @@ def execute_hook(base, options, block)
       end
     end
 
+    def already_run?(name, hook)
```
This method can be private

rafaelfranca commented on activesupport/lib/active_support/lazy_load_hooks.rb:
```diff
@@ -48,10 +63,20 @@ def execute_hook(base, options, block)
+      @only_once[name].include?(hook)
+    end
+
+    def flag_run(name, options, hook)
```
This method can be private

albertoalmagro commented:
@rafaelfranca done, I made methods private and rebased. Thanks!

kaspth commented on activesupport/lib/active_support/lazy_load_hooks.rb:
```diff
@@ -27,33 +27,60 @@ def self.extended(base) # :nodoc:
+        if options[:only_once]
+          @only_once[name] << hook
+        end
+      end
```
This seems a little too abstracted for my taste. I think it would be clearer if we kept the `once` checking responsibility within the same method:

```ruby
def execute_hook(name, base, options, block)
  with_once_execution(name, block, options[:only_once]) do
    if options[:yield]
      block.call(base)
    else
      base.instance_eval(&block)
    end
  end
end

def with_once_execution(name, block, once)
  unless @only_once[name].include?(block)
    @only_once[name] << block if once

    yield
  end
end
```

(I'd probably just call the option `once: true` since the `only_` is already implied.)

albertoalmagro commented:
@kaspth I really like your proposal, I have applied it but changing the method's name to `with_execution_control` instead of `with_once_execution`. `with_execution_control` suggests me that we are doing some control to determine if the hook will be executed or not whereas `with_once_execution` feels to me like it will always be one single execution.

I have also renamed the option to `:run_once` because I think it expresses better its intention.

@rafaelfranca please also review last changes.

rafaelfranca commented:
Just to confirm, does the application that you are reproducing this bug has `ActionController::Api` classes?

albertoalmagro commented:
Yes @rafaelfranca , I can confirm it. I have debugged it and I see how the hook is registered for both `ActionController::Base` and `ActionController::API`. I have also created a Rails app using this branch and reproduced the bug without using `:run_once`. After adding `:run_once` the problem is solved 😄

kaspth commented on activesupport/lib/active_support/lazy_load_hooks.rb:
```diff
@@ -40,20 +46,32 @@ def on_load(name, options = {}, &block)
+      end
+
+      def execute_hook(base, options, block)
+        with_execution_control(name, block, options[:run_once]) do
```
Where is `name` coming from here? Previously `execute_hook` was changed to pass it in.

albertoalmagro replied:
Good catch @kaspth I wonder why tests didn't blow up. Since this is merged, please tell me how should I fix it?

albertoalmagro replied:
@kaspth I already have a branch to pass the name. Should I open a new PR and reference this one?

rafaelfranca replied:
Yes please

albertoalmagro replied:
@rafaelfranca @kaspth Done please check PR #30285
